docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+
+The Tendermint mempool starts a new [broadcastTxRoutine][broadcast-tx-routine] for each peer that it is informed of. The routine sends all transactions that the mempool is aware of to all peers with few exceptions. The only exception is if the mempool received a transaction from a peer, then it marks it as such and won't resend to that peer. Otherwise, it retains no information about which transactions it already to sent to a peer. In some cases it may therefore resend transactions the peer already has. This can occur if the mempool removes a transaction from the `CList` data structure used to store the list of transaction while it is about to be sent and if the transaction was the tail of the `CList` during removal. This will be more likely to occur if a large number of transactions from the end of the list are removed during `RecheckTx`, since multiple transactions will become the tail and then be deleted. It is unclear at the moment how frequently this occurs on production chains.
+
+Beyond ensuring that transactions are rebroadcast to peers less frequently, there is not a simple scheme to communicate fewer transactions to peers. Peers cannot communicate what transactions they need since they do not know which transactions exist on the network.
+
+#### Vote Transmission
+
+Tendermint votes, both prevotes and precommits, are central to Tendermint consensus and are gossiped by all nodes to all peers during each consensus round. Data collected from the Blockpane node running on Osmosis indicates that the validator sounds about 9 gigabytes of `Vote` messages were sent during the nearly 3 hour period of observation.
+
+The Tendermint consensus reactor starts a new [gossipVotesRoutine][gossip-votes-routine] for each peer that it connects to. The reactor sends all votes to all peers unless it knows that the peer already has the vote or the reactor learns that the peer is in a different round and that thus the vote no longer applies. Tendermint learns that a peer has a vote in one of XX ways:
+
+1. Tendermint sent the peer the vote.
+2. Tendermint received the vote from the peer.
+3. The peer [sent a `HasVote` message][apply-has-vote]. This message is broadcast to all peers [each time validator receives a vote][publish-event-vote].
+4. The peer [sent a `VoteSetBits` message][apply-vote-set-bits]. This message is [sent as a response to a peer that sends a `VoteSetMaj23`][vote-set-bits-send].
+
+Given that Tendermint informs all peers of _each_ vote message it receives, all nodes should be well informed of which votes their peers have. Given that the vote messages were the third largest consumer of bandwidth in the observation on Osmosis, it's possible that this system is not currently working correctly. Further analysis should examine where votes may be being retransmitted.
+
+### Suggested Improvements to Lower Message Transmission Bandwidth
```

Comment:
Erasure coding block propagation is another one that (especially Dev) has been keen on. It decreases the chances you receive multiple parts because you only need, for instance, and 5 of 10 parts to reproduce the entire block. See https://github.com/tendermint/tendermint/issues/7932#issuecomment-913763248

Reply:
Erasure code will only help if we are more structured about how to spread information since each node still needs to receive the same amount of data for each proposal. 

For example, if the proposal data amounts to 40k or 10 "original" block parts and you add 5 blocks of "parity", then you have to spread 15 blocks from the proposer's point of view and each node needs to receive 10 block parts, original or parity. If some of the blocks are parity, then they will be used to reconstruct the missing original ones, also adding some processing overhead, before being able to reconstruct and ProcessProposal.